Make compaction manager switch to table abstraction

raphaelsc opened the pull request:
This work gets us a step closer to compaction groups. 

Everything in compaction layer but compaction_manager was converted to table_state. 

After this work, we can start implementing compaction groups, as each group will be represented by its own table_state. User-triggered operations that span the entire table, not only a group, can be done by calling the manager operation on behalf of each group and then merging the results, if any.

denesb commented:
> User-triggered operations that span the entire table, not only a group, can be done by calling the manager operation on behalf of each group and then merging the results, if any.

What are these operations? Scrub, cleanup, major? In case of the former, how is that working now?

raphaelsc commented:
> > User-triggered operations that span the entire table, not only a group, can be done by calling the manager operation on behalf of each group and then merging the results, if any.
> 
> What are these operations? Scrub, cleanup, major? In case of the former, how is that working now?

This series doesn't change semantics. Just continuing the work of abstracting table in compaction layer, compaction_manager was the only place left. Today there's a 1:1 relationship between table and table_state. But once we move to compaction groups, then 1 table can have multiple table_state, one for each group. I just mentioned it as a prediction of what we'll have to do once I push the initial compaction groups work.

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -827,8 +827,8 @@ void compaction_manager::do_stop() noexcept {
 
-inline bool compaction_manager::can_proceed(replica::table* t) const {
-    return (_state == state::enabled) && _compaction_state.contains(&t->as_table_state()) && !_compaction_state.at(&t->as_table_state()).compaction_disabled();
+inline bool compaction_manager::can_proceed(compaction::table_state* t) const {
```
Perhaps we should take the opportunity to switch to reference, AFAIK this pointer is never null.

bhalevy replied:
It is private to compaction_manager, and we use the pointer internally anyhow so converting back and forth between a pointer and a reference won't buy us anything.

denesb replied:
Well we are converting back-and-forth anyway because we have references in many entry points. I think if we are making changes we should moving towards references.

avikivity replied:
This series is long enough, I like the approach of just doing straightforward translations and not risking anything.

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -1006,7 +1006,7 @@ class compaction_manager::offstrategy_compaction_task : public compaction_manage
                 // at this moment reshape_candidates contains a set of sstables ready for integration into main set
                 auto desc = sstables::compaction_completion_desc{ .old_sstables = std::move(old_sstables), .new_sstables = std::move(reshape_candidates)};
-                co_await t.update_sstable_lists_on_off_strategy_completion(std::move(desc));
+                co_await t.as_table_state().on_compaction_completion(std::move(desc), sstables::offstrategy::yes);
```
This seems to also introduce a `_t.on_compaction_completion(std::move(desc));` that wasn't here before (through the new `on_compaction_completion()`). I don't know if this is a problem or not, but please clarify.

raphaelsc replied:
The idea is that we'll have a single on-completion interface for both "in-strategy" and off-strategy compactions, so not to pollute table_state with one interface for each. replica::table::on_compaction_completion is being moved into private namespace.

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -264,10 +264,10 @@ compaction_manager::compaction_state& compaction_manager::get_compaction_state(c
 }
 
-compaction_manager::task::task(compaction_manager& mgr, replica::table* t, sstables::compaction_type type, sstring desc)
+compaction_manager::task::task(compaction_manager& mgr, compaction::table_state* t, sstables::compaction_type type, sstring desc)
```
Again, we could switch to a reference here, if we are touching this code anyway.

denesb replied:
(also, the compaction:: qualifier seems extraneous but in no way harmful.

avikivity replied:
Let's defer this kind of changes. The series is long enough already.

denesb commented on compaction/compaction_manager.hh:
```diff
@@ -156,7 +156,7 @@ public:
         future<> run() noexcept;
 
-        const replica::table* compacting_table() const noexcept {
+        const compaction::table_state* compacting_table() const noexcept {
```
reference?

denesb commented on compaction/compaction_manager.hh:
```diff
@@ -413,17 +413,17 @@ public:
     class compaction_reenabler {
         compaction_manager& _cm;
-        replica::table* _table;
+        compaction::table_state* _table;
```
reference?

bhalevy replied:
The table may be removed in the background. We revalidate the pointer in submit(_table) when the compaction_reenabler is destroyed.

raphaelsc replied:
submit() will bail out earlier if it finds table_state is no longer registered, which happens once underlying table is destroyed.

bhalevy replied:
BTW, a new table object can be created with the same pointer and we wouldn't be able to distinguish between them.  A UUID would provide a more robust loose coupling mechanism...

avikivity replied:
Exactly.

But it's better to make remove() wait until there are no more instances of the table_state, rather let some stay alive and throw them away if they're invalid. The latter method is just waiting for a race to sneak in.

raphaelsc replied:
just noticed we're doing it correctly.   there's `gate::holder _holder` in compaction_reenabler, which we wait for in compaction_manager::remove().

avikivity replied:
It happens sometimes

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1477,6 +1477,10 @@ const std::vector<sstables::compaction_info> compaction_manager::get_compactions
 }
 
+bool compaction_manager::compaction_disabled(replica::table* t) const {
+    return _compaction_state.contains(&t->as_table_state()) && _compaction_state.at(&t->as_table_state()).compaction_disabled();
```
It's a waste to search the table state twice.
How about:
```c++
    auto it = _compaction_state.find(&t->as_table_state);
    return it != _compaction_state.end() && it->second.compaction_disabled();
```

raphaelsc replied:
I think we can do it after the series. As this series is very long, I decided to only proceed with straightforward switches, without much code change.

denesb commented on compaction/compaction_manager.hh:
```diff
@@ -393,7 +394,7 @@ public:
 
     // Submit a table to be upgraded and wait for its termination.
-    future<> perform_sstable_upgrade(replica::database& db, replica::table* t, bool exclude_current_version);
+    future<> perform_sstable_upgrade(replica::database& db, compaction::table_state& t, sstables::sstable_version_types last_version, bool exclude_current_version);
```
I thinks `s/last_version/version/` would be more fitting.

bhalevy replied:
We should be able to get to the table back from the table_state and through it get_highest_supported_format().
Adding it as a parameter is out of scope.
Also we may consider having a sstable instance per table_state / compaction_group as creation and deletion of sstables in it should be autonomous and therefore independent of the table.
(will become possibly once we switch to using uuid as the sstable generation - Cc @cmm)

avikivity replied:
I don't understand why we're adding a parameter.

raphaelsc replied:
> We should be able to get to the table back from the table_state and through it get_highest_supported_format(). Adding it as a parameter is out of scope. Also we may consider having a sstable instance per table_state / compaction_group as creation and deletion of sstables in it should be autonomous and therefore independent of the table. (will become possibly once we switch to using uuid as the sstable generation - Cc @cmm)

The idea of table_state is never referencing replica::table. But abstract table concept. table_state can be implemented without a replica::table at all, like for unit tests. 

I added a new parameter to avoid adding get_sstables_manager() to table_state interface, but I think it's perhaps better to just pass get_sstables_manager() to perform_sstable_upgrade().

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -1240,13 +1240,20 @@ class compaction_manager::validate_sstables_compaction_task : public compaction_
 
-future<> compaction_manager::perform_sstable_scrub_validate_mode(replica::table* t) {
+static std::vector<sstables::shared_sstable> get_all_sstables(compaction::table_state& t) {
+    auto s = boost::copy_range<std::vector<sstables::shared_sstable>>(*t.main_sstable_set().all());
```
I don't think you  need `boost::copy_range()` here. `std::vector` has a copy constructor, including one which operates on a pair of iterators.

denesb replied:
I see you copied this from below. I think this possibly had a filtering sometimes in the past, the filtering was removed and the `boost::copy_range()` survived.

raphaelsc replied:
You're spot on :-) will remove it. thanks

raphaelsc replied:
actually, turns out it's because sstable_set::all() 's return type is unordered_set.

bhalevy commented on compaction/compaction_manager.hh:
```diff
@@ -60,6 +60,9 @@ public:
     };
 private:
     struct compaction_state {
+        // FIXME: remove once we complete the transition to table_state.
```
We'll probably need a backpointer to the table, either from the table_state or from here when we have multiple table states / compaction groups per table.

raphaelsc replied:
I already did some initial work on compaction groups, and realized compaction_group class, which is stored in replica::table may need a reference to the underlying table. compaction_manager will never reference replica::table anymore from the moment this series is merged.

avikivity replied:
replica::table's subclass to compaction::compaction_group can have a backpointer. compaction::compaction_group should not know about tables.

denesb reviewed (approved):
LGTM, minor comments only.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1477,6 +1477,12 @@ const std::vector<sstables::compaction_info> compaction_manager::get_compactions
             }) | boost::adaptors::transformed(to_info));
 }
 
+bool compaction_manager::has_table_ongoing_compaction(const compaction::table_state& t) const {
```
I think that currently `compaction_manager` lives in the anonymous namespace.

bhalevy replied:
@raphaelsc can we drop the explicit `compaction::` namespace prefix?
compaction.hh has `using namespace compaction` so this shouldn't be needed.

Also, as a follow up, it looks like we should move more definition in compaction.hh and compaction_manager.hh into that namespace.

denesb commented:
> > > User-triggered operations that span the entire table, not only a group, can be done by calling the manager operation on behalf of each group and then merging the results, if any.
> > 
> > 
> > What are these operations? Scrub, cleanup, major? In case of the former, how is that working now?
> 
> This series doesn't change semantics. Just continuing the work of abstracting table in compaction layer, compaction_manager was the only place left. Today there's a 1:1 relationship between table and table_state. But once we move to compaction groups, then 1 table can have multiple table_state, one for each group. I just mentioned it as a prediction of what we'll have to do once I push the initial compaction groups work.

Yes, I wrote this after reading the cover letter but before reviewing the code.

bhalevy commented on compaction/table_state.hh:
```diff
@@ -28,7 +28,7 @@ public:
     virtual unsigned min_compaction_threshold() const noexcept = 0;
     virtual bool compaction_enforce_min_threshold() const noexcept = 0;
-    virtual const sstables::sstable_set& get_sstable_set() const = 0;
+    virtual const sstables::sstable_set& main_sstable_set() const = 0;
```
could be changed to `const noexcept` while at it.

avikivity replied:
Really, it adds nothing.

bhalevy commented on compaction/table_state.hh:
```diff
@@ -29,6 +29,7 @@ public:
     virtual unsigned min_compaction_threshold() const noexcept = 0;
     virtual bool compaction_enforce_min_threshold() const noexcept = 0;
     virtual const sstables::sstable_set& main_sstable_set() const = 0;
+    virtual const sstables::sstable_set& maintenance_sstable_set() const = 0;
```
const noexcept

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -191,6 +191,13 @@ void compaction_manager::deregister_weight(int weight) {
     reevaluate_postponed_compactions();
 }
 
+std::vector<sstables::shared_sstable> in_strategy_sstables(compaction::table_state& table_s) {
```
The patch title says `compaction:` but it is moved to compaction_manager.

I wonder if this series could be a good opportunity to also rename table_state to compaction_group.

denesb replied:
> I wonder if this series could be a good opportunity to also rename table_state to compaction_group.

I also thought of this. As we change the relationship of `table_state <-> table` from `1 <-> 1` to `N <-> 1`, the `compaction_group` name makes much more sense.

denesb replied:
Another (albeit minor) advantage is that it further detaches compaction from table, even if in name only.

bhalevy replied:
One more thing to consider for keeping table_state for now, though I don't like it and I'd like a better abstraction -
is TWCS double-decker design, where a compaction strategy is asked to compact a TWCS window during normal operation.  To make the two strategies composable (with STCS in master and ICS in enterprise), I think it would make sense to call `compaction_manager::submit` with an abstraction of the TWCS window.  This is sort of a compaction sub-group.  Today is a subset of the table's sstables and in the future it will be a subset of the compaction group sstables.
So this abstraction should map well to both cases and should point back to a compaction_group and the latter should point back to the table.

raphaelsc replied:
Each compaction group has a compaction policy which is inherited from the underlying replica::table. Alternatively, TWCS can just abstract the policy performed on each bucket, so upper layers don't have to care about it nor implement bucket isolation. With TWCS itself being aware of all buckets, cross-bucket compaction is simpler too, which happens with major.

bhalevy replied:
Cc @cmm

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -198,9 +198,9 @@ std::vector<sstables::shared_sstable> in_strategy_sstables(compaction::table_sta
 }
 
-std::vector<sstables::shared_sstable> compaction_manager::get_candidates(const replica::table& t) {
+std::vector<sstables::shared_sstable> compaction_manager::get_candidates(compaction::table_state& t) {
```
It's weird that table_state is in the compaction namespace.
Logically, it is an interface between the compaction_manager to table.
That's even more evident IMO when thinking about it as a compaction group.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1480,7 +1480,7 @@ const std::vector<sstables::compaction_info> compaction_manager::get_compactions
     using ret = std::vector<sstables::compaction_info>;
     return boost::copy_range<ret>(_tasks | boost::adaptors::filtered([t] (const shared_ptr<task>& task) {
-                return (!t || task->compacting_table() == t) && task->compaction_running();
+                return (!t || &task->compacting_table()->as_table_state() == t) && task->compaction_running();
```
No, that's really bad.
The task should work on a table_state (actually on a compaction group :))
and should expose it via a `task::compaction_group()` method that can be used here.

Note: with that I'm not sure we'll need `compacting_table()` anymore.
the compaction_group object should have a back reference to the table as compaction groups represent a subset of the table's data.

avikivity replied:
I think it will be cleared by following patches that change task::compacting_table() to return a table_state.

raphaelsc replied:
@avikiviy is right. I had to do this for making the series easier to read. I was only able to convert task much later due to cyclic dependencies.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1317/

bhalevy commented on api/compaction_manager.cc:
```diff
@@ -119,7 +119,7 @@ void set_compaction_manager(http_context& ctx, routes& r) {
             return parallel_for_each(table_names, [&db, &cm, &ks_name, type] (sstring& table_name) {
                 auto& t = db.find_column_family(ks_name, table_name);
-                return cm.stop_compaction(type, &t);
+                return cm.stop_compaction(type, &t.as_table_state());
```
I think that the compaction_manager can still expose `stop_compaction(table&)`
and we can have a variant of stop_ongoing_compactions that will still get a table*
and compare it the tasks' via each task compacting_table().

The per- table_state / compaction_group variant can stop ongoing tasks using
a respective task::compaction_group().

The reason is we don't have to expose the complexity to the caller, and force the callers (api, table) to iterate through the table compaction groups to stop each one of them - requiring a full traversal of the compaction_manager tasks for each.  We better traverse it once and look up the table.

raphaelsc replied:
Yes, initially I thought of adding a stop_compaction() variant that accepts ks_name and cf_name, so APIs can use it to stop tasks running on a replica::table. But then I realized that mixing replica::table and compaction::table_state would bring lots of complexity to compaction_manager. By limiting interaction with manager to table_state, we have a more solid interface.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1125,7 +1125,6 @@ class compaction_manager::rewrite_sstables_compaction_task : public compaction_m
 
         for (;;) {
             switch_state(state::active);
-            replica::table& t = *_compacting_table;
```
It looks like it is already unused in master.
Please submit separately.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1005,7 +1005,8 @@ class compaction_manager::offstrategy_compaction_task : public compaction_manage
             if (desc.sstables.empty()) {
                 // at this moment reshape_candidates contains a set of sstables ready for integration into main set
-                co_await t.update_sstable_lists_on_off_strategy_completion(old_sstables, reshape_candidates);
+                auto desc = sstables::compaction_completion_desc{ .old_sstables = std::move(old_sstables), .new_sstables = std::move(reshape_candidates)};
```
let's make it multi-line so the designated initializers become more visible
(the space character between the opening curly and `.old_sstables` hurts my eyes :-)

raphaelsc replied:
Will change. thanks!

bhalevy commented on replica/table.cc:
```diff
@@ -856,18 +856,17 @@ table::sstable_list_builder::build_new_list(const sstables::sstable_set& current
     public:
-        explicit sstable_lists_updater(table& t, sstable_list_builder::permit_t permit, const sstables_t& old_maintenance, const sstables_t& new_main)
+        explicit sstable_lists_updater(table& t, sstable_list_builder::permit_t permit, const sstables_t old_maintenance, const sstables_t new_main)
                 : _t(t), _builder(std::move(permit)), _old_maintenance(old_maintenance), _new_main(new_main) {
```
If we copy them, why pass by value?
If passing by value, why not move the args?

raphaelsc replied:
you're right. will move.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -330,7 +330,7 @@ future<sstables::compaction_result> compaction_manager::task::compact_sstables(s
             release_exhausted(desc.old_sstables);
         }
-        t.on_compaction_completion(std::move(desc));
+        t.as_table_state().on_compaction_completion(std::move(desc), sstables::offstrategy::no).get();
```
The task should work on the table_state / a compaction_group in the first place.
It shouldn't even know that a table has (currently) a `as_table_state` method.

denesb replied:
This is all removed later.You should have a look at the final state in the changed files tab.

bhalevy replied:
So maybe the patch series needs to go top down, using a back reference from table_state to table where needed, and removing those as the internal interfaces are converted to accepting table_state.

denesb replied:
Is a complete reorganization of 30 patches worth it to avoid a transient layering violation?

bhalevy replied:
IMO, yes.  We better add shims where need downstream, rather than temporarily change the api just for reverting it later.
It makes the series harder to reason about and to test.

bhalevy commented on compaction/table_state.hh:
```diff
@@ -40,6 +40,7 @@ public:
     virtual api::timestamp_type min_memtable_timestamp() const = 0;
     virtual future<> update_compaction_history(utils::UUID compaction_id, sstring ks_name, sstring cf_name, std::chrono::milliseconds ended_at, int64_t bytes_in, int64_t bytes_out) = 0;
     virtual future<> on_compaction_completion(sstables::compaction_completion_desc desc, sstables::offstrategy offstrategy) = 0;
+    virtual bool is_auto_compaction_disabled_by_user() const = 0;
```
const noexcept please

raphaelsc replied:
will add.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1428,7 +1425,7 @@ future<> compaction_manager::perform_sstable_scrub(replica::table* t, sstables::
 }
 
 void compaction_manager::add(replica::table* t) {
```
we'll need to add / remove individual compaction_group:s
as we'd want to split / merge them dynamically.

The bottom line is that the compaction_manager api should switch almost entirely to using table_state / compaction_group (plus the ability to stop compaction by table since it's a common use case that currently requires O(n) traversal of all all tasks)

avikivity replied:
In the end compation/ shouldn't include anything from replica/.

raphaelsc replied:
That's what we'll achieve by the end of this series.

bhalevy commented on api/storage_service.cc:
```diff
@@ -1409,7 +1409,7 @@ void set_snapshot(http_context& ctx, routes& r, sharded<db::snapshot_ctl>& snap_
                     auto& cm = db.get_compaction_manager();
                     auto& cf = db.find_column_family(keyspace, cfname);
-                    return cm.perform_sstable_scrub(&cf, opts);
+                    return cm.perform_sstable_scrub(cf.as_table_state(), opts);
```
I think we should add a table method to do that, rather than open coding `as_table_state` here and in other layers above table.

When the table will have multiple compaction group it will need to scrub all of them anyhow (and synchronize with compaction group changes).

avikivity replied:
I agree, but I think it's best deferred to a later series and let this one focus on separating compaction/ and replica/.

raphaelsc replied:
Yes, when implementing this, it became clear that for multiple groups, we'll need replica::table being a bridge to all of its groups, so if api/storage_service wants to perform an operation which spans all groups, then replica::table will call manager on behalf of all groups, and only then return to the caller. And yes, the idea is that this series is the first step towards group. once this series is in, then we'll need to do work on table for supporting multiple groups.

bhalevy commented on test/boost/view_complex_test.cc:
```diff
@@ -840,7 +840,7 @@ void test_commutative_row_deletion(cql_test_env& e, std::function<void()>&& mayb
     });
 
-    e.local_db().get_compaction_manager().perform_major_compaction(&e.local_db().find_column_family("ks", "vcf")).get();
+    e.local_db().get_compaction_manager().perform_major_compaction(e.local_db().find_column_family("ks", "vcf").as_table_state()).get();
```
why can't the test use table::compact_all_sstables()?

bhalevy replied:
it's not testing the compaction_manager...

raphaelsc replied:
Looks like an improvement for later. Tried to minimize the changes needed as much as possible.

avikivity commented on compaction/compaction_manager.cc:
```diff
@@ -326,11 +326,11 @@ future<sstables::compaction_result> compaction_manager::task::compact_sstables(s
         if (release_exhausted) {
             release_exhausted(desc.old_sstables);
         }
+        t.on_compaction_completion(std::move(desc));
```
Can this change in order cause problems?

avikivity replied:
It's good practice to make these refactoring have as little effect as possible, and push out more dangerous changes out.

avikivity commented:
Looks good. I think we should try to limit the scope of the series to just disentangling compaction from replica::table and not add additional cleanups.

raphaelsc commented:
> Missing: patch that removes replica/ includes from compaction/.

Forgot the trophy :-)

raphaelsc commented:
> Looks good. I think we should try to limit the scope of the series to just disentangling compaction from replica::table and not add additional cleanups.

These cleanups can be moved into a preparatory series:
"table: use compaction_completion_desc to describe changes for off-strategy
table: remove ref from on_compaction_completion() signature
compaction_manager: remove unused variable in rewrite_sstable()"

It seems all the other patches will have to go here.

raphaelsc commented:
v2:
- extracted cleanup patches into a preparatory series that was already merged
- upgrade: get highest supported format through table_state, to avoid adding a new param.
- added noexcept qualifier to is_auto_compaction_disabled_by_user().
- remove forward declaration of replica::table in compaction_manager.hh

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1387/

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1491,8 +1491,8 @@ const std::vector<sstables::compaction_info> compaction_manager::get_compactions
 
-bool compaction_manager::compaction_disabled(replica::table* t) const {
-    return _compaction_state.contains(&t->as_table_state()) && _compaction_state.at(&t->as_table_state()).compaction_disabled();
+bool compaction_manager::compaction_disabled(compaction::table_state& t) const {
```
Please change to const reference.  This function is not meant to change the table state.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -340,10 +340,10 @@ future<sstables::compaction_result> compaction_manager::task::compact_sstables(s
     co_return co_await sstables::compact_sstables(std::move(descriptor), cdata, t.as_table_state());
 }
-future<> compaction_manager::task::update_history(replica::table& t, const sstables::compaction_result& res, const sstables::compaction_data& cdata) {
+future<> compaction_manager::task::update_history(compaction::table_state& t, const sstables::compaction_result& res, const sstables::compaction_data& cdata) {
```
the `compaction::` namespace prefix is probably not needed as compaction.hh is `using namespace compaction`

bhalevy commented on test/boost/cql_query_large_test.cc:
```diff
@@ -117,7 +117,7 @@ SEASTAR_THREAD_TEST_CASE(test_large_data) {
         e.db().invoke_on_all([] (replica::database& dbi) {
             return parallel_for_each(dbi.get_column_families(), [&dbi] (auto& table) {
-                return dbi.get_compaction_manager().perform_major_compaction(&*table.second);
+                return dbi.get_compaction_manager().perform_major_compaction((table.second)->as_table_state());
```
nit: did you forget to remove the parenthesis?

raphaelsc replied:
yes. sorry.

bhalevy reviewed (commented):
Looks good.  I had only minor comments.
